File: backend/src/main/java/com/wms/service/ReturnsService.java
package com.wms.service;

import com.wms.dto.request.*;
import com.wms.dto.response.CreditNoteResponse;
import com.wms.dto.response.ReceiptActionResponse;
import com.wms.entity.*;
import com.wms.enums.*;
import com.wms.exception.BusinessRuleViolationException;
import com.wms.exception.ResourceNotFoundException;
import com.wms.repository.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class ReturnsService {

    private static final Logger log = LoggerFactory.getLogger(ReturnsService.class);

    private static final String RECEIPT_ENTITY = "RECEIPT";
    private static final String INVENTORY_ENTITY = "INVENTORY";
    private static final String DEALER_ENTITY = "DEALER";
    private static final String CREDIT_NOTE_ENTITY = "CREDIT_NOTE";

    private final ReceiptRepository receiptRepository;
    private final ReceiptItemRepository receiptItemRepository;
    private final DeliveryOrderRepository deliveryOrderRepository;
    private final DeliveryOrderItemRepository deliveryOrderItemRepository;
    private final DealerRepository dealerRepository;
    private final WarehouseLocationRepository warehouseLocationRepository;
    private final InventoryRepository inventoryRepository;
    private final CreditNoteRepository creditNoteRepository;
    private final ReceiptValidationService receiptValidationService;
    private final AuditLogService auditLogService;

    public ReturnsService(ReceiptRepository receiptRepository,
                          ReceiptItemRepository receiptItemRepository,
                          DeliveryOrderRepository deliveryOrderRepository,
                          DeliveryOrderItemRepository deliveryOrderItemRepository,
                          DealerRepository dealerRepository,
                          WarehouseLocationRepository warehouseLocationRepository,
                          InventoryRepository inventoryRepository,
                          CreditNoteRepository creditNoteRepository,
                          ReceiptValidationService receiptValidationService,
                          AuditLogService auditLogService) {
        this.receiptRepository = receiptRepository;
        this.receiptItemRepository = receiptItemRepository;
        this.deliveryOrderRepository = deliveryOrderRepository;
        this.deliveryOrderItemRepository = deliveryOrderItemRepository;
        this.dealerRepository = dealerRepository;
        this.warehouseLocationRepository = warehouseLocationRepository;
        this.inventoryRepository = inventoryRepository;
        this.creditNoteRepository = creditNoteRepository;
        this.receiptValidationService = receiptValidationService;
        this.auditLogService = auditLogService;
    }

    @Transactional
    public ReceiptActionResponse createReturnReceipt(CreateReturnRequest request, User actor) {
        receiptValidationService.assertRole(actor, UserRole.WAREHOUSE_STAFF, "RETURN_RECEIPT_CREATE");
        receiptValidationService.assertWarehouseAccess(actor, request.getWarehouseId());

        DeliveryOrder doOrder = deliveryOrderRepository.findById(request.getDeliveryOrderId())
                .orElseThrow(() -> new ResourceNotFoundException("Delivery Order not found: " + request.getDeliveryOrderId()));

        if (!doOrder.getDealer().getId().equals(request.getDealerId())) {
            throw new BusinessRuleViolationException(
                    "DEALER_MISMATCH: Returned dealer does not match original Delivery Order dealer");
        }

        if (doOrder.getStatus() != DeliveryOrderStatus.DELIVERED && doOrder.getStatus() != DeliveryOrderStatus.COMPLETED) {
            throw new BusinessRuleViolationException(
                    "INVALID_DO_STATUS: Returns are only allowed for DELIVERED or COMPLETED Delivery Orders. Current status: " + doOrder.getStatus());
        }

        List<DeliveryOrderItem> doItems = deliveryOrderItemRepository.findByDeliveryOrderId(doOrder.getId());
        List<Receipt> existingReturns = receiptRepository.findByDeliveryOrderIdAndType(doOrder.getId(), ReceiptType.RETURN);

        List<ReceiptItem> receiptItemsToSave = new ArrayList<>();

        // Validate each returned product against DO sales limits
        for (CreateReturnItemRequest itemReq : request.getItems()) {
            final Long reqProductId = itemReq.getProductId();
            DeliveryOrderItem matchingDoItem = doItems.stream()
                    .filter(doi -> doi.getProduct().getId().equals(reqProductId))
                    .findFirst()
                    .orElseThrow(() -> new BusinessRuleViolationException(
                            "PRODUCT_NOT_IN_DO: Product ID " + reqProductId + " is not in the original Delivery Order"));

            BigDecimal previouslyReturnedQty = BigDecimal.ZERO;
            for (Receipt returnReceipt : existingReturns) {
                List<ReceiptItem> items = receiptItemRepository.findByReceiptId(returnReceipt.getId());
                BigDecimal sum = items.stream()
                        .filter(ri -> ri.getProduct().getId().equals(reqProductId))
                        .map(ri -> BigDecimal.valueOf(ri.getExpectedQty()))
                        .reduce(BigDecimal.ZERO, BigDecimal::add);
                previouslyReturnedQty = previouslyReturnedQty.add(sum);
            }

            BigDecimal currentReturnQty = BigDecimal.valueOf(itemReq.getExpectedQty());
            BigDecimal totalReturned = previouslyReturnedQty.add(currentReturnQty);
            BigDecimal issuedQty = matchingDoItem.getIssuedQty();

            if (totalReturned.compareTo(issuedQty) > 0) {
                throw new BusinessRuleViolationException(
                        "RETURN_EXCEEDS_ORIGINAL_SALE: Total returned quantity (" + totalReturned 
                        + ") exceeds original sales quantity (" + issuedQty + ") for product ID " + reqProductId);
            }

            ReceiptItem receiptItem = ReceiptItem.builder()
                    .product(matchingDoItem.getProduct())
                    .batch(matchingDoItem.getBatch()) // Retain original batch for FIFO tracking
                    .expectedQty(itemReq.getExpectedQty())
                    .unitCost(matchingDoItem.getUnitPrice()) // Use original unit price for credit calculation
                    .qcResult(QcResult.PENDING)
                    .overReceivedQty(0)
                    .build();

            receiptItemsToSave.add(receiptItem);
        }

        // Generate Return Receipt Number
        String receiptNumber = "REC-RET-" + LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd"))
                + "-" + UUID.randomUUID().toString().substring(0, 6).toUpperCase();

        Receipt receipt = Receipt.builder()
                .receiptNumber(receiptNumber)
                .sourceOrderCode(doOrder.getDoNumber())
                .type(ReceiptType.RETURN)
                .warehouse(doOrder.getWarehouse())
                .dealer(doOrder.getDealer())
                .deliveryOrder(doOrder)
                .status(ReceiptStatus.DRAFT)
                .documentDate(LocalDate.now())
                .accountingPeriod(doOrder.getAccountingPeriod())
                .createdBy(actor)
                .notes(request.getNotes())
                .createdAt(OffsetDateTime.now())
                .updatedAt(OffsetDateTime.now())
                .version(1)
                .build();

        receipt = receiptRepository.save(receipt);

        for (ReceiptItem item : receiptItemsToSave) {
            item.setReceipt(receipt);
            receiptItemRepository.save(item);
        }

        auditLogService.log(
                actor, AuditAction.RECEIPT_CREATE, RECEIPT_ENTITY,
                receipt.getId(), receipt.getReceiptNumber(),
                receipt.getWarehouse().getId(),
                Map.of(),
                Map.of("type", ReceiptType.RETURN.name(), "status", ReceiptStatus.DRAFT.name())
        );

        return ReceiptActionResponse.builder()
                .id(receipt.getId())
                .receiptNumber(receipt.getReceiptNumber())
                .status(receipt.getStatus())
                .version(receipt.getVersion())
                .updatedAt(receipt.getUpdatedAt())
                .message("Tạo phiếu trả hàng thành công")
                .build();
    }

    @Transactional
    public ReceiptActionResponse processReturnQc(Long receiptId, ReturnQcRequest request, User actor) {
        receiptValidationService.assertRole(actor, UserRole.WAREHOUSE_STAFF, "RETURN_QC");
        
        Receipt receipt = receiptRepository.findById(receiptId)
                .orElseThrow(() -> new ResourceNotFoundException("Receipt not found: " + receiptId));

        if (receipt.getType() != ReceiptType.RETURN) {
            throw new BusinessRuleViolationException("INVALID_RECEIPT_TYPE: Receipt is not a RETURN receipt");
        }

        if (receipt.getStatus() != ReceiptStatus.DRAFT) {
            throw new BusinessRuleViolationException(
                    "INVALID_STATE: QC processing can only be executed in DRAFT state. Current status: " + receipt.getStatus());
        }

        receiptValidationService.assertWarehouseAccess(actor, receipt.getWarehouse().getId());
        receiptValidationService.assertVersionMatch(receipt, request.getExpectedVersion());

        List<DeliveryOrderItem> doItems = deliveryOrderItemRepository.findByDeliveryOrderId(receipt.getDeliveryOrder().getId());

        for (ReturnQcItemRequest itemReq : request.getItems()) {
            final ReceiptItem item = receiptItemRepository.findById(itemReq.getReceiptItemId())
                    .orElseThrow(() -> new ResourceNotFoundException("Receipt item not found: " + itemReq.getReceiptItemId()));

            if (!item.getReceipt().getId().equals(receiptId)) {
                throw new BusinessRuleViolationException("ITEM_RECEIPT_MISMATCH: Item does not belong to receipt " + receiptId);
            }

            if (itemReq.getPassedQty() + itemReq.getFailedQty() != itemReq.getActualQty()) {
                throw new BusinessRuleViolationException("QC_SAMPLE_SUM_MISMATCH: passedQty + failedQty must equal actualQty");
            }

            final Long productId = item.getProduct().getId();
            final DeliveryOrderItem matchingDoItem = doItems.stream()
                    .filter(doi -> doi.getProduct().getId().equals(productId))
                    .findFirst()
                    .orElseThrow(() -> new BusinessRuleViolationException("PRODUCT_NOT_IN_DO"));

            // 1. Process Passed Qty (Increase Regular Inventory)
            if (itemReq.getPassedQty() > 0) {
                WarehouseLocation passedLoc = warehouseLocationRepository.findById(itemReq.getPassedLocationId())
                        .orElseThrow(() -> new ResourceNotFoundException("Location not found: " + itemReq.getPassedLocationId()));

                if (Boolean.TRUE.equals(passedLoc.getIsQuarantine())) {
                    throw new BusinessRuleViolationException(
                            "INVALID_LOCATION: Passed target must be a regular (non-quarantine) Bin. Location: " + itemReq.getPassedLocationId());
                }

                // Assert capacity
                BigDecimal addedVol = BigDecimal.valueOf(itemReq.getPassedQty()).multiply(
                        item.getProduct().getVolumeM3() != null ? item.getProduct().getVolumeM3() : BigDecimal.ZERO);
                BigDecimal addedWt = BigDecimal.valueOf(itemReq.getPassedQty()).multiply(
                        item.getProduct().getWeightKg() != null ? item.getProduct().getWeightKg() : BigDecimal.ZERO);

                BigDecimal curVol = passedLoc.getCurrentVolumeM3() != null ? passedLoc.getCurrentVolumeM3() : BigDecimal.ZERO;
                BigDecimal curWt = passedLoc.getCurrentWeightKg() != null ? passedLoc.getCurrentWeightKg() : BigDecimal.ZERO;

                if (passedLoc.getCapacityM3() != null && curVol.add(addedVol).compareTo(passedLoc.getCapacityM3()) > 0) {
                    throw new BusinessRuleViolationException("BIN_CAPACITY_EXCEEDED: Passed volume exceeds location capacity");
                }
                if (passedLoc.getCapacityKg() != null && curWt.add(addedWt).compareTo(passedLoc.getCapacityKg()) > 0) {
                    throw new BusinessRuleViolationException("BIN_CAPACITY_EXCEEDED: Passed weight exceeds location capacity");
                }

                final Warehouse finalWarehouse = receipt.getWarehouse();
                final Product finalProduct = item.getProduct();
                final Batch finalBatch = item.getBatch();
                final WarehouseLocation finalPassedLoc = passedLoc;
                final BigDecimal finalUnitCost = matchingDoItem.getUnitCost();

                // Update inventory
                Inventory inventory = inventoryRepository
                        .findByWarehouseProductBatchLocationForUpdate(
                                finalWarehouse.getId(),
                                finalProduct.getId(),
                                finalBatch.getId(),
                                finalPassedLoc.getId())
                        .orElseGet(() -> Inventory.builder()
                                .warehouse(finalWarehouse)
                                .product(finalProduct)
                                .batch(finalBatch)
                                .location(finalPassedLoc)
                                .totalQty(BigDecimal.ZERO)
                                .reservedQty(BigDecimal.ZERO)
                                .costPrice(finalUnitCost != null ? finalUnitCost : BigDecimal.ZERO) // keep original cost price
                                .updatedAt(OffsetDateTime.now())
                                .build());

                BigDecimal oldQty = inventory.getTotalQty();
                inventory.setTotalQty(oldQty.add(BigDecimal.valueOf(itemReq.getPassedQty())));
                inventory.setUpdatedAt(OffsetDateTime.now());
                inventoryRepository.save(inventory);

                // Update location occupancy
                passedLoc.setCurrentVolumeM3(curVol.add(addedVol));
                passedLoc.setCurrentWeightKg(curWt.add(addedWt));
                passedLoc.setUpdatedAt(OffsetDateTime.now());
                warehouseLocationRepository.save(passedLoc);

                // Audit Log
                auditLogService.log(
                        actor, AuditAction.INVENTORY_UPDATE, INVENTORY_ENTITY,
                        inventory.getId(), "INV-" + receipt.getWarehouse().getId() + "-" + item.getProduct().getId(),
                        receipt.getWarehouse().getId(),
                        Map.of("totalQty", oldQty),
                        Map.of("totalQty", inventory.getTotalQty(), "locationId", passedLoc.getId(), "delta", itemReq.getPassedQty())
                );
            }

            // 2. Process Failed Qty (Increase Quarantine Inventory)
            if (itemReq.getFailedQty() > 0) {
                WarehouseLocation quarantineLoc = warehouseLocationRepository.findById(itemReq.getQuarantineLocationId())
                        .orElseThrow(() -> new ResourceNotFoundException("Location not found: " + itemReq.getQuarantineLocationId()));

                if (!Boolean.TRUE.equals(quarantineLoc.getIsQuarantine())) {
                    throw new BusinessRuleViolationException(
                            "INVALID_LOCATION: Failed target must be a quarantine location. Location: " + itemReq.getQuarantineLocationId());
                }

                // Assert capacity
                BigDecimal addedVol = BigDecimal.valueOf(itemReq.getFailedQty()).multiply(
                        item.getProduct().getVolumeM3() != null ? item.getProduct().getVolumeM3() : BigDecimal.ZERO);
                BigDecimal addedWt = BigDecimal.valueOf(itemReq.getFailedQty()).multiply(
                        item.getProduct().getWeightKg() != null ? item.getProduct().getWeightKg() : BigDecimal.ZERO);

                BigDecimal curVol = quarantineLoc.getCurrentVolumeM3() != null ? quarantineLoc.getCurrentVolumeM3() : BigDecimal.ZERO;
                BigDecimal curWt = quarantineLoc.getCurrentWeightKg() != null ? quarantineLoc.getCurrentWeightKg() : BigDecimal.ZERO;

                if (quarantineLoc.getCapacityM3() != null && curVol.add(addedVol).compareTo(quarantineLoc.getCapacityM3()) > 0) {
                    throw new BusinessRuleViolationException("BIN_CAPACITY_EXCEEDED: Quarantine volume exceeds location capacity");
                }
                if (quarantineLoc.getCapacityKg() != null && curWt.add(addedWt).compareTo(quarantineLoc.getCapacityKg()) > 0) {
                    throw new BusinessRuleViolationException("BIN_CAPACITY_EXCEEDED: Quarantine weight exceeds location capacity");
                }

                final Warehouse finalWarehouse = receipt.getWarehouse();
                final Product finalProduct = item.getProduct();
                final Batch finalBatch = item.getBatch();
                final WarehouseLocation finalQuarantineLoc = quarantineLoc;
                final BigDecimal finalUnitCost = matchingDoItem.getUnitCost();

                // Update inventory
                Inventory inventory = inventoryRepository
                        .findByWarehouseProductBatchLocationForUpdate(
                                finalWarehouse.getId(),
                                finalProduct.getId(),
                                finalBatch.getId(),
                                finalQuarantineLoc.getId())
                        .orElseGet(() -> Inventory.builder()
                                .warehouse(finalWarehouse)
                                .product(finalProduct)
                                .batch(finalBatch)
                                .location(finalQuarantineLoc)
                                .totalQty(BigDecimal.ZERO)
                                .reservedQty(BigDecimal.ZERO)
                                .costPrice(finalUnitCost != null ? finalUnitCost : BigDecimal.ZERO) // keep original cost price
                                .updatedAt(OffsetDateTime.now())
                                .build());

                BigDecimal oldQty = inventory.getTotalQty();
                inventory.setTotalQty(oldQty.add(BigDecimal.valueOf(itemReq.getFailedQty())));
                inventory.setUpdatedAt(OffsetDateTime.now());
                inventoryRepository.save(inventory);

                // Update location occupancy
                quarantineLoc.setCurrentVolumeM3(curVol.add(addedVol));
                quarantineLoc.setCurrentWeightKg(curWt.add(addedWt));
                quarantineLoc.setUpdatedAt(OffsetDateTime.now());
                warehouseLocationRepository.save(quarantineLoc);

                // Audit Log
                auditLogService.log(
                        actor, AuditAction.INVENTORY_UPDATE, INVENTORY_ENTITY,
                        inventory.getId(), "INV-" + receipt.getWarehouse().getId() + "-" + item.getProduct().getId(),
                        receipt.getWarehouse().getId(),
                        Map.of("totalQty", oldQty),
                        Map.of("totalQty", inventory.getTotalQty(), "locationId", quarantineLoc.getId(), "delta", itemReq.getFailedQty())
                );
            }

            // 3. Update ReceiptItem QC status
            item.setActualQty(itemReq.getActualQty());
            item.setSampleQty(itemReq.getActualQty());
            item.setSamplePassedQty(itemReq.getPassedQty());
            item.setSampleFailedQty(itemReq.getFailedQty());
            item.setQcResult(itemReq.getFailedQty() > 0 ? QcResult.FAILED : QcResult.PASSED);
            item.setLocation(warehouseLocationRepository.findById(itemReq.getPassedLocationId()).orElse(null));
            item.setQcBy(actor);
            receiptItemRepository.save(item);
        }

        // Set receipt status directly to APPROVED
        receipt.setStatus(ReceiptStatus.APPROVED);
        receipt.setApprovedBy(actor);
        receipt.setApprovedAt(OffsetDateTime.now());
        receipt.setUpdatedAt(OffsetDateTime.now());
        receipt = receiptRepository.save(receipt);

        auditLogService.log(
                actor, AuditAction.RECEIPT_APPROVE, RECEIPT_ENTITY,
                receipt.getId(), receipt.getReceiptNumber(),
                receipt.getWarehouse().getId(),
                Map.of("status", ReceiptStatus.DRAFT.name()),
                Map.of("status", ReceiptStatus.APPROVED.name(), "approvedBy", actor.getId(), "approvedAt", receipt.getApprovedAt().toString())
        );

        return ReceiptActionResponse.builder()
                .id(receipt.getId())
                .receiptNumber(receipt.getReceiptNumber())
                .status(receipt.getStatus())
                .version(receipt.getVersion())
                .updatedAt(receipt.getUpdatedAt())
                .message("Đã hoàn tất phân tách QC và duyệt nhập kho hàng trả")
                .build();
    }

    @Transactional
    public CreditNoteResponse createCreditNote(Long receiptId, CreateCreditNoteRequest request, User actor) {
        // Assert role WAREHOUSE_MANAGER, ACCOUNTANT (includes ACCOUNTANT_MANAGER, CEO, ADMIN)
        if (actor.getRole() != UserRole.ACCOUNTANT && actor.getRole() != UserRole.ACCOUNTANT_MANAGER 
                && actor.getRole() != UserRole.WAREHOUSE_MANAGER && actor.getRole() != UserRole.CEO && actor.getRole() != UserRole.ADMIN) {
            throw new BusinessRuleViolationException("FORBIDDEN_CREDIT_NOTE_ROLE: User is not authorized to generate credit notes");
        }

        Receipt receipt = receiptRepository.findById(receiptId)
                .orElseThrow(() -> new ResourceNotFoundException("Receipt not found: " + receiptId));

        if (receipt.getType() != ReceiptType.RETURN) {
            throw new BusinessRuleViolationException("INVALID_RECEIPT_TYPE: Credit note can only be created for RETURN receipts");
        }

        if (receipt.getStatus() != ReceiptStatus.APPROVED) {
            throw new BusinessRuleViolationException("INVALID_STATE: Credit note requires APPROVED return receipt status");
        }

        // Check if credit note already generated
        if (creditNoteRepository.findByReceiptId(receiptId).isPresent()) {
            throw new BusinessRuleViolationException("CREDIT_NOTE_ALREADY_EXISTS: Credit note has already been generated for this receipt");
        }

        receiptValidationService.assertWarehouseAccess(actor, receipt.getWarehouse().getId());

        List<ReceiptItem> items = receiptItemRepository.findByReceiptId(receiptId);
        BigDecimal totalRefundAmount = BigDecimal.ZERO;

        for (ReceiptItem item : items) {
            BigDecimal qty = item.getActualQty() != null ? BigDecimal.valueOf(item.getActualQty()) : BigDecimal.ZERO;
            BigDecimal returnPrice = item.getUnitCost() != null ? item.getUnitCost() : BigDecimal.ZERO; // unitCost contains the original DO unitPrice
            totalRefundAmount = totalRefundAmount.add(qty.multiply(returnPrice));
        }

        Dealer dealer = receipt.getDealer();
        BigDecimal oldBalance = dealer.getCurrentBalance();
        BigDecimal newBalance = oldBalance.subtract(totalRefundAmount); // Refund reduces outstanding balance dealer owes us

        dealer.setCurrentBalance(newBalance);
        dealerRepository.save(dealer);

        // Generate Credit Note Number
        String creditNoteNumber = "CN-" + LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd"))
                + "-" + UUID.randomUUID().toString().substring(0, 6).toUpperCase();

        CreditNote creditNote = CreditNote.builder()
                .creditNoteNumber(creditNoteNumber)
                .dealer(dealer)
                .receipt(receipt)
                .amount(totalRefundAmount)
                .reason(request.getReason())
                .createdBy(actor)
                .documentDate(LocalDate.now())
                .accountingPeriod(receipt.getAccountingPeriod())
                .createdAt(OffsetDateTime.now())
                .build();

        creditNote = creditNoteRepository.save(creditNote);

        auditLogService.log(
                actor, AuditAction.CREDIT_NOTE_CREATE, CREDIT_NOTE_ENTITY,
                creditNote.getId(), creditNoteNumber,
                receipt.getWarehouse().getId(),
                Map.of(),
                Map.of("amount", totalRefundAmount, "dealerId", dealer.getId())
        );

        auditLogService.log(
                actor, AuditAction.UPDATE, DEALER_ENTITY,
                dealer.getId(), dealer.getCode(),
                receipt.getWarehouse().getId(),
                Map.of("balance", oldBalance),
                Map.of("balance", newBalance, "delta", totalRefundAmount.negate(), "creditNoteId", creditNote.getId())
        );

        return CreditNoteResponse.builder()
                .creditNoteId(creditNote.getId())
                .creditNoteNumber(creditNoteNumber)
                .dealerId(dealer.getId())
                .dealerName(dealer.getName())
                .amount(totalRefundAmount)
                .currentBalance(newBalance)
                .reason(request.getReason())
                .documentDate(LocalDate.now())
                .createdAt(OffsetDateTime.now())
                .build();
    }
}

File: frontend/src/services/returns.service.js
import apiClient, { useMock } from './api.client';

const BASE = '/returns';

// Helpers for mock local storage database
const KEYS = {
  RECEIPTS: 'wms_db_receipts',
  RECEIPT_ITEMS: 'wms_db_receipt_items',
  DELIVERY_ORDERS: 'wms_db_delivery_orders',
  DELIVERY_ORDER_ITEMS: 'wms_db_delivery_order_items',
  DEALERS: 'wms_db_dealers',
  CREDIT_NOTES: 'wms_db_credit_notes',
  INVENTORIES: 'wms_db_inventories',
  LOCATIONS: 'wms_db_warehouse_locations'
};

function getDb(key, fallback = []) {
  try {
    const val = localStorage.getItem(key);
    return val ? JSON.parse(val) : fallback;
  } catch (e) {
    return fallback;
  }
}

function saveDb(key, data) {
  try {
    localStorage.setItem(key, JSON.stringify(data));
  } catch (e) {
    console.error(e);
  }
}

const returnsService = {
  async getReturns(params = {}) {
    if (useMock) {
      await delay(300);
      const receipts = getDb(KEYS.RECEIPTS, []);
      let list = receipts.filter(r => r.type === 'RETURN');
      if (params.warehouse_id) {
        list = list.filter(r => r.warehouse_id === Number(params.warehouse_id));
      }
      return list;
    }
    const response = await apiClient.get('/receipts', {
      params: { ...params, type: 'RETURN' }
    });
    return response.data;
  },

  async getReturnById(id) {
    if (useMock) {
      await delay(200);
      const receipts = getDb(KEYS.RECEIPTS, []);
      const returnReceipt = receipts.find(r => r.id === Number(id));
      if (!returnReceipt) throw new Error('RETURN_RECEIPT_NOT_FOUND');

      const items = getDb(KEYS.RECEIPT_ITEMS, []).filter(ri => ri.receipt_id === returnReceipt.id);
      
      return {
        ...returnReceipt,
        items
      };
    }
    const response = await apiClient.get(`/receipts/${id}`);
    return response.data;
  },

  async createReturn(data) {
    if (useMock) {
      await delay(500);
      const receipts = getDb(KEYS.RECEIPTS, []);
      const receiptItems = getDb(KEYS.RECEIPT_ITEMS, []);
      const dos = getDb(KEYS.DELIVERY_ORDERS, []);
      const doItems = getDb(KEYS.DELIVERY_ORDER_ITEMS, []);
      const dealers = getDb(KEYS.DEALERS, []);

      const doOrder = dos.find(d => d.id === Number(data.deliveryOrderId));
      if (!doOrder) throw new Error('DELIVERY_ORDER_NOT_FOUND');

      const dealer = dealers.find(dl => dl.id === Number(data.dealerId));
      if (!dealer) throw new Error('DEALER_NOT_FOUND');

      if (doOrder.dealer_id !== dealer.id) {
        throw new Error('DEALER_MISMATCH');
      }

      const receiptId = receipts.length > 0 ? Math.max(...receipts.map(r => r.id)) + 1 : 1;
      const receiptNumber = `REC-RET-${new Date().toISOString().slice(0, 10).replace(/-/g, '')}-${Math.random().toString(36).substring(2, 8).toUpperCase()}`;

      const newReceipt = {
        id: receiptId,
        receipt_number: receiptNumber,
        source_order_code: doOrder.do_number,
        type: 'RETURN',
        warehouse_id: Number(data.warehouseId),
        dealer_id: dealer.id,
        dealer_name: dealer.name,
        delivery_order_id: doOrder.id,
        status: 'DRAFT',
        document_date: new Date().toISOString().slice(0, 10),
        notes: data.notes || '',
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
        version: 1
      };

      const newItems = data.items.map((itemReq, index) => {
        const doItem = doItems.find(doi => doi.do_id === doOrder.id && doi.product_id === Number(itemReq.productId));
        if (!doItem) throw new Error(`PRODUCT_NOT_IN_DO: Product ${itemReq.productId}`);

        return {
          id: receiptItems.length + index + 1,
          receipt_id: receiptId,
          product_id: Number(itemReq.productId),
          product_sku: doItem.product_sku || `SKU-${itemReq.productId}`,
          product_name: doItem.product_name || `Product Name ${itemReq.productId}`,
          expected_qty: Number(itemReq.expectedQty),
          actual_qty: null,
          sample_qty: null,
          sample_passed_qty: null,
          sample_failed_qty: null,
          qc_result: 'PENDING',
          unit_cost: doItem.unit_price, // store original sales price for credit calculation
          batch_id: doItem.batch_id || 1
        };
      });

      receipts.push(newReceipt);
      saveDb(KEYS.RECEIPTS, receipts);
      saveDb(KEYS.RECEIPT_ITEMS, [...receiptItems, ...newItems]);

      return {
        id: receiptId,
        receiptNumber,
        status: 'DRAFT',
        version: 1,
        message: 'Tạo phiếu trả hàng thành công (Mock)'
      };
    }
    const response = await apiClient.post(BASE, data);
    return response.data;
  },

  async processQc(id, data) {
    if (useMock) {
      await delay(600);
      const receipts = getDb(KEYS.RECEIPTS, []);
      const receiptItems = getDb(KEYS.RECEIPT_ITEMS, []);
      const inventories = getDb(KEYS.INVENTORIES, []);
      const locations = getDb(KEYS.LOCATIONS, []);

      const rIdx = receipts.findIndex(r => r.id === Number(id));
      if (rIdx === -1) throw new Error('RECEIPT_NOT_FOUND');

      const receipt = receipts[rIdx];
      if (receipt.status !== 'DRAFT') throw new Error('RECEIPT_NOT_IN_DRAFT');

      data.items.forEach(qcReq => {
        const riIdx = receiptItems.findIndex(ri => ri.id === Number(qcReq.receiptItemId));
        if (riIdx === -1) return;

        const item = receiptItems[riIdx];
        item.actual_qty = Number(qcReq.actualQty);
        item.sample_qty = Number(qcReq.actualQty);
        item.sample_passed_qty = Number(qcReq.passedQty);
        item.sample_failed_qty = Number(qcReq.failedQty);
        item.qc_result = qcReq.failedQty > 0 ? 'FAILED' : 'PASSED';
        item.location_id = Number(qcReq.passedLocationId);

        // Update inventories for passed quantity
        if (qcReq.passedQty > 0) {
          const passLoc = locations.find(l => l.id === Number(qcReq.passedLocationId));
          const invIdx = inventories.findIndex(inv => 
            inv.warehouse_id === receipt.warehouse_id &&
            inv.product_id === item.product_id &&
            inv.location_id === Number(qcReq.passedLocationId)
          );

          if (invIdx !== -1) {
            inventories[invIdx].total_qty = Number(inventories[invIdx].total_qty) + qcReq.passedQty;
            inventories[invIdx].updated_at = new Date().toISOString();
          } else {
            inventories.push({
              id: inventories.length + 1,
              warehouse_id: receipt.warehouse_id,
              product_id: item.product_id,
              batch_id: item.batch_id || 1,
              location_id: Number(qcReq.passedLocationId),
              total_qty: qcReq.passedQty,
              reserved_qty: 0,
              cost_price: 85000,
              updated_at: new Date().toISOString()
            });
          }

          if (passLoc) {
            passLoc.current_volume_m3 = Number((passLoc.current_volume_m3 + qcReq.passedQty * 0.05).toFixed(3));
            passLoc.current_weight_kg = Number((passLoc.current_weight_kg + qcReq.passedQty * 1.5).toFixed(2));
          }
        }

        // Update inventories for failed quantity (quarantine)
        if (qcReq.failedQty > 0) {
          const failLoc = locations.find(l => l.id === Number(qcReq.quarantineLocationId));
          const invIdx = inventories.findIndex(inv => 
            inv.warehouse_id === receipt.warehouse_id &&
            inv.product_id === item.product_id &&
            inv.location_id === Number(qcReq.quarantineLocationId)
          );

          if (invIdx !== -1) {
            inventories[invIdx].total_qty = Number(inventories[invIdx].total_qty) + qcReq.failedQty;
            inventories[invIdx].updated_at = new Date().toISOString();
          } else {
            inventories.push({
              id: inventories.length + 1,
              warehouse_id: receipt.warehouse_id,
              product_id: item.product_id,
              batch_id: item.batch_id || 1,
              location_id: Number(qcReq.quarantineLocationId),
              total_qty: qcReq.failedQty,
              reserved_qty: 0,
              cost_price: 85000,
              updated_at: new Date().toISOString()
            });
          }

          if (failLoc) {
            failLoc.current_volume_m3 = Number((failLoc.current_volume_m3 + qcReq.failedQty * 0.05).toFixed(3));
            failLoc.current_weight_kg = Number((failLoc.current_weight_kg + qcReq.failedQty * 1.5).toFixed(2));
          }
        }
      });

      receipt.status = 'APPROVED';
      receipt.approved_at = new Date().toISOString();
      receipt.version += 1;
      receipts[rIdx] = receipt;

      saveDb(KEYS.RECEIPTS, receipts);
      saveDb(KEYS.RECEIPT_ITEMS, receiptItems);
      saveDb(KEYS.INVENTORIES, inventories);
      saveDb(KEYS.LOCATIONS, locations);

      return {
        id: receipt.id,
        receiptNumber: receipt.receipt_number,
        status: 'APPROVED',
        version: receipt.version,
        updatedAt: receipt.approved_at,
        message: 'Đã hoàn tất phân tách QC và nhập kho hàng trả (Mock)'
      };
    }
    const response = await apiClient.put(`${BASE}/${id}/qc`, data);
    return response.data;
  },

  async createCreditNote(id, data) {
    if (useMock) {
      await delay(500);
      const receipts = getDb(KEYS.RECEIPTS, []);
      const receiptItems = getDb(KEYS.RECEIPT_ITEMS, []);
      const dealers = getDb(KEYS.DEALERS, []);
      const creditNotes = getDb(KEYS.CREDIT_NOTES, []);

      const receipt = receipts.find(r => r.id === Number(id));
      if (!receipt) throw new Error('RECEIPT_NOT_FOUND');

      const dlIdx = dealers.findIndex(d => d.id === receipt.dealer_id);
      if (dlIdx === -1) throw new Error('DEALER_NOT_FOUND');

      const items = receiptItems.filter(ri => ri.receipt_id === receipt.id);
      const totalAmount = items.reduce((sum, item) => sum + (item.actual_qty || 0) * (item.unit_cost || 0), 0);

      const dealer = dealers[dlIdx];
      const oldBalance = dealer.current_balance;
      dealer.current_balance = oldBalance - totalAmount;
      dealers[dlIdx] = dealer;

      const cnId = creditNotes.length + 1;
      const creditNoteNumber = `CN-${new Date().toISOString().slice(0, 10).replace(/-/g, '')}-${Math.random().toString(36).substring(2, 8).toUpperCase()}`;

      const newCreditNote = {
        creditNoteId: cnId,
        creditNoteNumber,
        dealerId: dealer.id,
        dealerName: dealer.name,
        amount: totalAmount,
        currentBalance: dealer.current_balance,
        reason: data.reason,
        documentDate: new Date().toISOString().slice(0, 10),
        createdAt: new Date().toISOString()
      };

      creditNotes.push(newCreditNote);
      
      // Update receipt to know credit note exists in mock database
      receipt.credit_note_generated = true;
      const rIdx = receipts.findIndex(r => r.id === receipt.id);
      if (rIdx !== -1) receipts[rIdx] = receipt;

      saveDb(KEYS.RECEIPTS, receipts);
      saveDb(KEYS.DEALERS, dealers);
      saveDb(KEYS.CREDIT_NOTES, creditNotes);

      return newCreditNote;
    }
    const response = await apiClient.post(`${BASE}/${id}/credit-note`, data);
    return response.data;
  }
};

function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export default returnsService;
